### PR TITLE
[search] SearchAPI tests.

### PR DESCRIPTION
**Pull request by ygorshenin**
TestSearchEngine is decoupled with Index, so it's possible to abstract out base test that allows to build custom mwms, and reuse it in search integration tests and in search API tests.

### CODE REVIEW

File: search/search_integration_tests/helpers.cpp [hunk 1]
```diff
@@ -1,44 +1,22 @@
...
-                        MercatorBounds::maxX, MercatorBounds::maxY));
-}
-
-SearchTest::~SearchTest()
+  : m_scopedLog(LDEBUG)
+  , m_engine(m_index, make_unique<storage::CountryInfoGetterForTesting>(), Engine::Params{})
 {
-  for (auto const & file : m_files)
-    Cleanup(file);
+  SetViewport(m2::RectD(MercatorBounds::minX, MercatorBounds::minY, MercatorBounds::maxX,
```
Comment (mpimenov) [inline search/search_integration_tests/helpers.cpp]: MercatorBounds::FullRect()

File: search/search_integration_tests/helpers.cpp [hunk 2]
```diff
@@ -108,14 +86,18 @@ size_t SearchTest::CountFeatures(m2::RectD const & rect)
...
   size_t count = 0;
   auto counter = [&count](const FeatureType & /* ft */) { ++count; };
-  m_engine.ForEachInRect(counter, rect, scales::GetUpperScale());
+  m_index.ForEachInRect(counter, rect, scales::GetUpperScale());
   return count;
 }
 
 // static
-void SearchTest::Cleanup(platform::LocalCountryFile const & map)
+void SearchTest::OnMwmBuilded(MwmInfo const & info)
```
Comment (mpimenov) [inline search/search_integration_tests/helpers.cpp]: s/Builded/Built/

File: map/map_tests/search_api_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
+// NOTE: the purpose of this test is to test interaction between
+// SearchAPI and search engine. If you would like to test search
+// engine behaviour, please, consider to implement one more search
```
[the remark this answers is not in the document]
Comment (ygorshenin) [reply map/map_tests/search_api_tests.cpp]: Done

File: map/map_tests/search_api_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+protected:
+  Storage m_storage;
+  unique_ptr<CountryInfoGetter> m_infoGetter;
+  Delegate m_delegate;
+  SearchAPI m_api;
+};
+
+UNIT_CLASS_TEST(SearchAPITest, MultipleViewportRequests)
```
[the remark this answers is not in the document]
Comment (ygorshenin) [reply map/map_tests/search_api_tests.cpp]: Done

File: search/search_integration_tests/helpers.cpp (hunk 1 above)
Comment (ygorshenin) [reply search/search_integration_tests/helpers.cpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @VladiMihaylenko PTAL

File: map/map_tests/search_api_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    if (stage == 0)
+    {
+      Rules const rules = {ExactMatch(id, cafe1), ExactMatch(id, cafe2)};
+      TEST(MatchResults(m_index, rules, results), ());
+
+      promise0.set_value();
+    }
+    else
+    {
+      TEST_EQUAL(stage, 1, ());
```
Comment (mpimenov) [inline map/map_tests/search_api_tests.cpp]: s/TEST/CHECK/ ?

Comment (ygorshenin) [reply map/map_tests/search_api_tests.cpp]: Probably, in tests we need to use TEST macros.

I usually prefer CHECKs in helper classes/functions (as they may be re-used in tools), but in a test body I prefer to use TEST macros.

**Pull request merged**